pkg/batchmapper/service.go

```diff
@@ -0,0 +1,126 @@
+	// that were received on the stream. We use an atomic int as this needs to be synchronized
+	// between the request/response go routines.
+	totalRequests := atomic.NewInt32(0)
+
+	// datumStreamCh is used to stream messages to the user code interface
+	// As the BatchMap interface expects a list of request elements
+	// we read all the requests coming on the stream and keep streaming them to the user code on this channel.
+	datumStreamCh := make(chan Datum)
+
+	// go routine to invoke the user handler function, and process the responses.
+	g.Go(func() error {
+		// Apply the user BatchMap implementation function
+		responses := fs.BatchMapper.BatchMap(ctx, datumStreamCh)
+
+		// If the number of responses received does not align with the request batch size,
+		// we will not be able to process the data correctly.
+		// This should be marked as an error and the container is restarted.
+		// As this is a user error, we restart the container to mitigate any transient error otherwise, this
+		// crash should indicate to the user that there is some issue.
+		if len(responses.Items()) != int(totalRequests.Load()) {
+			errMsg := fmt.Sprintf("batchMapFn: mismatch between length of batch requests and responses, "+
+				"expected:%d, got:%d", int(totalRequests.Load()), len(responses.Items()))
+			log.Panic(errMsg)
+		}
+
+		// iterate over the responses received and covert to the required proto format
+		for _, batchResp := range responses.Items() {
+			var elements []*batchmappb.BatchMapResponse_Result
+			for _, resp := range batchResp.Items() {
+				elements = append(elements, &batchmappb.BatchMapResponse_Result{
+					Keys:  resp.Keys(),
+					Value: resp.Value(),
+					Tags:  resp.Tags(),
+				})
+			}
+			singleRequestResp := &batchmappb.BatchMapResponse{
+				Results: elements,
+				Id:      batchResp.Id(),
+			}
+			// We stream back the result for a single request ID
+			// this would contain all the responses for that request.
+			err := stream.Send(singleRequestResp)
+			if err != nil {
+				log.Println("BatchMapFn: Got an error while Send() on stream", err)
+				return err
+			}
+		}
+		return nil
+	})
+
+	// loop to keep reading messages from the stream and sending it to the datumStreamCh
+	for {
+		d, err := stream.Recv()
+		// if we see EOF on the stream we do not have any more messages coming up
+		if err == io.EOF {
+			// close the input data channel to indicate that no more messages expected
+			close(datumStreamCh)
+			break
+		}
+		if err != nil {
+			// close the input data channel to indicate that no more messages expected
+			close(datumStreamCh)
+			log.Println("BatchMapFn: Got an error while recv() on stream", err)
+			return err
+		}
+		var hd = NewHandlerDatum(d.GetValue(), d.GetEventTime().AsTime(), d.GetWatermark().AsTime(), d.GetHeaders(), d.GetId(), d.GetKeys())
+		// send the datum to the input channel
+		datumStreamCh <- hd
+		// Increase the counter for number of requests received
+		totalRequests.Inc()
```

Comment:
I preferred to update this once we are sure that the message has been sent to the channel, ie consumed by the user code. So if they return early, at that point we are sure about the correct count of number of requests sent to them.
Any things that come to your mind which can cause an issue?

Reply:
Sorry, I forgot to mention my concern. I was looking at this piece of code:

```
if err != nil {
			// close the input data channel to indicate that no more messages expected
			close(datumStreamCh)
			log.Println("BatchMapFn: Got an error while recv() on stream", err)
			return err
		}
```

Since the error is not EOF, I wonder if we still receive a datum. If so, with our current flow, the request count is not increased. Hence we received a datum but didn't get to send it to the channel. On the collecting response side, the count still matches, would it cause a data loss?

We are returning an error, will it fail the entire batch? Please confirm. Thanks!

Reply:
In case of an error we would have return and fail + replay the entire batch